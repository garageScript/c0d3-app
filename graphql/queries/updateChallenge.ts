import { gql } from 'apollo-boost'

const UPDATE_CHALLENGE = gql`
  mutation updateChallenge(
    $lessonId: Int!
    $order: Int
    $description: String
    $title: String
    $id: Int!
  ) {
    updateChallenge(
      id: $id
      lessonId: $lessonId
      order: $order
      description: $description
      title: $title
    ) {
      id
      docUrl
      githubUrl
      videoUrl
      chatUrl
      order
      description
      title
      challenges {
        id
        description
        lessonId
        title
        order
      }
    }
  }
`

export default UPDATE_CHALLENGE
