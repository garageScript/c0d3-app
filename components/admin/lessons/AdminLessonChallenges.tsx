import { useMutation } from '@apollo/client'
import React, { useEffect, useState } from 'react'
import updateChallenge from '../../../graphql/queries/updateChallenge'
import { FormCard } from '../../FormCard'
import createNewChallenge from '../../../graphql/queries/createChallenge'
import { Lesson, Challenge, Maybe } from '../../../graphql/index'
import {
  getPropertyArr,
  makeGraphqlVariable,
  errorCheckAllFields
} from '../../../helpers/admin/adminHelpers'
import { lessonSchema } from '../../../helpers/formValidation'
import { formChange } from '../../../helpers/formChange'

const challengeAttributes = {
  title: '',
  description: '',
  order: ''
}

type LessonChallengeProps = {
  challenge: Maybe<Challenge>
  alter: (options: any) => Promise<void>
}

type NewChallengeProps = {
  setLessons: React.Dispatch<React.SetStateAction<Lesson[] | null>>
  lessonId: number
}
type LessonChallengesProps = {
  setLessons: React.Dispatch<React.SetStateAction<Lesson[] | null>>
  challenges: Maybe<Challenge>[] | null | undefined
  lessonId?: number
}

// Renders when someone clicks on `create new button` on the sidebar
export const NewChallenge: React.FC<NewChallengeProps> = ({
  setLessons,
  lessonId
}) => {
  const [createChallenge, { loading, data }] = useMutation(createNewChallenge)
  const [challengeProperties, setChallengeProperties] = useState(
    getPropertyArr(challengeAttributes)
  )
  // when data is fully loaded after sending mutation request, update front-end lessons info
  useEffect(() => {
    !loading && data && setLessons(data.createChallenge)
  }, [data])

  // alter gets called when someone clicks button to create a lesson
  const alter = async () => {
    const newProperties = [...challengeProperties]
    const valid = await errorCheckAllFields(newProperties, lessonSchema)
    if (!valid) {
      setChallengeProperties(newProperties)
      return
    }

    try {
      await createChallenge(
        makeGraphqlVariable(challengeProperties, { lessonId })
      )

      window.location.reload()
    } catch (err) {
      throw new Error(err)
    }
  }

  const handleChange = async (value: string, propertyIndex: number) => {
    await formChange(
      value,
      propertyIndex,
      challengeProperties,
      setChallengeProperties,
      lessonSchema
    )
  }

  return (
    <div className="mb-2">
      <div className="mb-3 text-center">
        <span className="text-primary font-weight-bold display-3">
          Create New Challenge
        </span>
      </div>

      <FormCard
        onChange={handleChange}
        values={challengeProperties}
        onSubmit={{ title: 'Create Challenge', onClick: alter }}
      />
    </div>
  )
}

const LessonChallenge: React.FC<LessonChallengeProps> = ({
  challenge,
  alter
}) => {
  const [challengeProperties, setChallengeProperties] = useState(
    getPropertyArr(challenge, ['lessonId', '__typename'])
  )

  const handleChange = async (value: string, propertyIndex: number) => {
    await formChange(
      value,
      propertyIndex,
      challengeProperties,
      setChallengeProperties,
      lessonSchema
    )
  }

  const handleSubmit = {
    title: 'Update Challenge',
    onClick: async () => {
      const newProperties = [...challengeProperties]
      const valid = await errorCheckAllFields(newProperties, lessonSchema)
      if (!valid) {
        setChallengeProperties(newProperties)
        return
      }
      alter(challengeProperties)
    }
  }

  return (
    <div className="mt-3">
      <FormCard
        onChange={handleChange}
        title={(challenge && challenge.title + '') || ''}
        values={challengeProperties}
        onSubmit={handleSubmit}
      />
    </div>
  )
}

// creates list of cards to update challenges
export const AdminLessonChallenges: React.FC<LessonChallengesProps> = ({
  setLessons,
  challenges,
  lessonId
}) => {
  const [alterChallenge, { loading, data }] = useMutation(updateChallenge)

  // when data is fully loaded after sending mutation request, update front-end lessons info
  useEffect(() => {
    !loading && data && setLessons(data.updateChallenge)
  }, [data])

  // alter gets called when someone clicks button to update a challenge
  const alter = async (options: any) => {
    try {
      await alterChallenge(makeGraphqlVariable(options, { lessonId }))
      window.location.reload()
    } catch (err) {
      throw new Error(err)
    }
  }

  const allChallenges = !challenges
    ? []
    : challenges.map((challenge: Maybe<Challenge>, key: number) => (
        <LessonChallenge challenge={challenge} alter={alter} key={key} />
      ))

  return <>{allChallenges}</>
}
