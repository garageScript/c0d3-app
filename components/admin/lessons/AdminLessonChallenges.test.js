import '../../../__mocks__/useIsMac.mock'
import '../../../__mocks__/useBreakpoint.mock'
import * as React from 'react'
import { render, screen, waitFor } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import * as Sentry from '@sentry/browser'
import { MockedProvider } from '@apollo/client/testing'
import { AdminLessonChallenges, NewChallenge } from './AdminLessonChallenges'
import dummyLessonData from '../../../__dummy__/lessonData'
import updateChallenge from '../../../graphql/queries/updateChallenge'
import createNewChallenge from '../../../graphql/queries/createChallenge'
jest.spyOn(Sentry, 'captureException').mockImplementation(() => {})
const challenges = [
  {
    id: '107',
    title: 'Sum of 2 Numbers',
    description:
      "Write a function that takes in 2 numbers and returns their sum.\n\nHere's how another developer might use your function:\n\n```\nsolution(5,9) // Should return 14\nsolution(4,1) // Should return 5\n```",
    order: 1
  },
  {
    id: '108',
    title: 'Sum of 3 Numbers',
    description:
      "Write a function that takes in 3 numbers and returns their sum.\n\nHere's how another developer might use your function:\n\n```\nsolution(5,9,2) // Should return 16\nsolution(4,1,9) // Should return 14\n```",
    order: 2
  }
]
const updateChallengeMock = {
  request: {
    query: updateChallenge,
    variables: {
      id: 107,
      title: 'New title',
      description: 'New description',
      order: 15,
      lessonId: 0
    }
  },
  result: {
    data: { updateChallenge: dummyLessonData[0] }
  }
}
const createNewChallengeMock = {
  request: {
    query: createNewChallenge,
    variables: {
      title: 'New challenge',
      description: 'New challenge description',
      order: 2,
      lessonId: 0
    }
  },
  result: {
    data: { createChallenge: dummyLessonData[0] }
  }
}

const mocks = [updateChallengeMock, createNewChallengeMock]

describe('AdminLessonChallenges component', () => {
  delete window.location
  const reload = jest.fn()
  window.location = { reload }
  beforeEach(() => {
    jest.clearAllMocks()
  })
  test('Should update challenge', async () => {
    const { container } = render(
      <MockedProvider mocks={mocks} addTypename={false}>
        <AdminLessonChallenges
          challenges={challenges}
          lessonId={0}
          setLessons={() => true}
        />
      </MockedProvider>
    )
    expect(container).toMatchSnapshot()
    const title = screen.getAllByTestId('input1')[0]
    const description = screen.getAllByTestId('textbox')[0]
    const order = screen.getAllByTestId('input3')[0]

    await userEvent.type(title, '{backspace}New title', {
      delay: 1,
      initialSelectionEnd: title.value.length,
      initialSelectionStart: 0
    })
    await userEvent.type(description, '{backspace}New description', {
      delay: 1,
      initialSelectionEnd: description.value.length,
      initialSelectionStart: 0
    })
    await userEvent.type(order, '{backspace}15', {
      delay: 1,
      initialSelectionEnd: order.value.length,
      initialSelectionStart: 0
    })
    await waitFor(() =>
      userEvent.click(screen.getAllByText('Update Challenge')[0])
    )
    await waitFor(() => expect(reload).toBeCalled())
    await waitFor(() => expect(container).toMatchSnapshot())
  })
  test('Should create challenge', async () => {
    const { container } = render(
      <MockedProvider mocks={mocks} addTypename={false}>
        <NewChallenge
          challenges={challenges}
          lessonId={0}
          setLessons={() => true}
        />
      </MockedProvider>
    )
    const title = screen.getByTestId('input0')
    const description = screen.getByTestId('textbox')
    const order = screen.getByTestId('input2')
    await userEvent.type(title, 'New challenge', { delay: 1 })
    await userEvent.type(description, 'New challenge description', { delay: 1 })
    await userEvent.type(order, '2', { delay: 1 })
    await waitFor(() => userEvent.click(screen.getByText('Create Challenge')))
    await waitFor(() => expect(reload).toBeCalledTimes(1))
    await waitFor(() => expect(container).toMatchSnapshot())
  })
  test('Should refuse updating challenge with incorrect info', async () => {
    const { container } = render(
      <MockedProvider mocks={mocks} addTypename={false}>
        <AdminLessonChallenges
          challenges={challenges}
          lessonId={0}
          setLessons={() => true}
        />
      </MockedProvider>
    )
    const title = screen.getAllByTestId('input1')[0]
    const description = screen.getAllByTestId('textbox')[0]
    const order = screen.getAllByTestId('input3')[0]
    title.setSelectionRange(0, title.value.length)
    description.setSelectionRange(0, description.value.length)
    order.setSelectionRange(0, order.value.length)
    await userEvent.type(title, '{backspace}New title', { delay: 1 })
    await userEvent.type(description, '{backspace}New description', {
      delay: 1
    })
    await userEvent.type(order, '{backspace}', { delay: 1 })
    await waitFor(() =>
      userEvent.click(screen.getAllByText('Update Challenge')[0])
    )
    await waitFor(() => expect(reload).not.toBeCalled())
    await waitFor(() => expect(screen.getByText('Required')).toBeTruthy())
    expect(container).toMatchSnapshot()
  })
  test('Should refuse creating  new challenge with incorrect info', async () => {
    const { container } = render(
      <MockedProvider mocks={mocks} addTypename={false}>
        <NewChallenge
          challenges={challenges}
          lessonId={0}
          setLessons={() => true}
        />
      </MockedProvider>
    )
    const title = screen.getByTestId('input0')
    const description = screen.getByTestId('textbox')
    const order = screen.getByTestId('input2')
    await userEvent.type(title, 'New challenge', { delay: 1 })
    await userEvent.type(description, 'New challenge description', { delay: 1 })
    await userEvent.type(order, '{backspace}', { delay: 1 })
    await waitFor(() => userEvent.click(screen.getByText('Create Challenge')))
    await waitFor(() => expect(reload).not.toBeCalled())
    await waitFor(() => expect(screen.getByText('Required')).toBeTruthy())
    await waitFor(() => expect(container).toMatchSnapshot())
  })
  test('Should capture error when updating', async () => {
    const errorMock = { ...updateChallengeMock, error: new Error('fail') }
    render(
      <MockedProvider mocks={[errorMock]} addTypename={false}>
        <AdminLessonChallenges
          challenges={challenges}
          lessonId={0}
          setLessons={() => true}
        />
      </MockedProvider>
    )
    const titles = screen.getAllByTestId('input1')
    const descriptions = screen.getAllByTestId('textbox')
    const orders = screen.getAllByTestId('input3')
    await userEvent.clear(titles[0])
    await userEvent.clear(descriptions[0])
    await userEvent.clear(orders[0])
    await userEvent.type(titles[0], 'New title', { delay: 1 })
    await userEvent.type(descriptions[0], 'New description', { delay: 1 })
    await userEvent.type(orders[0], '15', { delay: 1 })
    await waitFor(() =>
      userEvent.click(screen.getAllByText('Update Challenge')[0])
    )
    await waitFor(() => expect(Sentry.captureException).toBeCalled())
  })
  test('Should capture error when creating new challenge', async () => {
    const errorMock = { ...createNewChallengeMock, error: new Error('fail') }
    render(
      <MockedProvider mocks={[errorMock]} addTypename={false}>
        <NewChallenge
          challenges={challenges}
          lessonId={0}
          setLessons={() => true}
        />
      </MockedProvider>
    )
    const title = screen.getByTestId('input0')
    const description = screen.getByTestId('textbox')
    const order = screen.getByTestId('input2')
    await userEvent.type(title, 'New challenge', { delay: 1 })
    await userEvent.type(description, 'New challenge description', { delay: 1 })
    await userEvent.type(order, '2', { delay: 1 })
    await waitFor(() => userEvent.click(screen.getByText('Create Challenge')))
    await waitFor(() => expect(Sentry.captureException).toBeCalled())
  })
  test('Should render undefined challenges', async () => {
    const { container } = render(
      <MockedProvider mocks={mocks} addTypename={false}>
        <AdminLessonChallenges lessonId={0} setLessons={() => true} />
      </MockedProvider>
    )
    expect(container).toMatchSnapshot()
  })
  test('Should render nulled challenges', async () => {
    const { container } = render(
      <MockedProvider mocks={mocks} addTypename={false}>
        <AdminLessonChallenges
          lessonId={0}
          challenges={[null, null]}
          setLessons={() => true}
        />
      </MockedProvider>
    )
    expect(container).toMatchSnapshot()
  })
  test('Should render a challenge without title', async () => {
    const titlessChallenge = { ...challenges[0] }
    delete titlessChallenge.title
    const { container } = render(
      <MockedProvider mocks={mocks} addTypename={false}>
        <AdminLessonChallenges
          challenges={[titlessChallenge]}
          lessonId={0}
          setLessons={() => true}
        />
      </MockedProvider>
    )
    await waitFor(() => expect(container).toMatchSnapshot())
  })
})
