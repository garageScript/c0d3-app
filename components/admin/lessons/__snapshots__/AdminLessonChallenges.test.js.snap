// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AdminLessonChallenges component Should create challenge 1`] = `
<div>
  <div
    class="mb-2"
  >
    <div
      class="mb-3 text-center"
    >
      <span
        class="text-primary font-weight-bold display-3"
      >
        Create New Challenge
      </span>
    </div>
    <div
      class="row"
    >
      <div
        class="card shadow-sm col-12"
      >
        <div
          class="card-body text-center"
        />
        <div
          class="text-left"
        >
          <div
            class="d-flex flex-column ml-3 mr-3 mb-4"
          >
            <h5
              data-testid="h5title0"
            >
              Title
            </h5>
            <input
              class="form-control"
              data-testid="input0"
              placeholder=""
              type="text"
              value="New challenge"
            />
          </div>
          <div
            class="d-flex flex-column ml-3 mr-3 mb-4"
          >
            <h5
              data-testid="h5description1"
            >
              Description
            </h5>
            <div
              style="background-color: white;"
            >
              <button
                class="btn"
                style="color: rgb(41, 41, 41);"
              >
                Write
              </button>
              <button
                class="btn"
                style="color: rgb(142, 142, 142);"
              >
                Preview
              </button>
              <textarea
                data-testid="textbox"
                placeholder="Type something..."
                style="padding: 1rem; width: 100%;"
              >
                New challenge description
              </textarea>
            </div>
          </div>
          <div
            class="d-flex flex-column ml-3 mr-3 mb-4"
          >
            <h5
              data-testid="h5order2"
            >
              Order
            </h5>
            <input
              class="form-control"
              data-testid="input2"
              placeholder=""
              type="text"
              value="2"
            />
          </div>
        </div>
        <div
          class="text-center mb-4"
        >
          <button
            class="btn bg-primary"
            style="color: rgb(255, 255, 255);"
          >
            Create Challenge
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`AdminLessonChallenges component Should refuse creating  new challenge with incorrect info 1`] = `
<div>
  <div
    class="mb-2"
  >
    <div
      class="mb-3 text-center"
    >
      <span
        class="text-primary font-weight-bold display-3"
      >
        Create New Challenge
      </span>
    </div>
    <div
      class="row"
    >
      <div
        class="card shadow-sm col-12"
      >
        <div
          class="card-body text-center"
        />
        <div
          class="text-left"
        >
          <div
            class="d-flex flex-column ml-3 mr-3 mb-4"
          >
            <h5
              data-testid="h5title0"
            >
              Title
            </h5>
            <input
              class="form-control"
              data-testid="input0"
              placeholder=""
              type="text"
              value="New challenge"
            />
          </div>
          <div
            class="d-flex flex-column ml-3 mr-3 mb-4"
          >
            <h5
              data-testid="h5description1"
            >
              Description
            </h5>
            <div
              style="background-color: white;"
            >
              <button
                class="btn"
                style="color: rgb(41, 41, 41);"
              >
                Write
              </button>
              <button
                class="btn"
                style="color: rgb(142, 142, 142);"
              >
                Preview
              </button>
              <textarea
                data-testid="textbox"
                placeholder="Type something..."
                style="padding: 1rem; width: 100%;"
              >
                New challenge description
              </textarea>
            </div>
          </div>
          <div
            class="d-flex flex-column ml-3 mr-3 mb-4"
          >
            <h5
              data-testid="h5order2"
            >
              Order
            </h5>
            <input
              class="form-control"
              data-testid="input2"
              placeholder=""
              type="text"
              value=""
            />
            <h6
              class="text-danger mb-0"
            >
              Required
            </h6>
          </div>
        </div>
        <div
          class="text-center mb-4"
        >
          <button
            class="btn bg-primary"
            style="color: rgb(255, 255, 255);"
          >
            Create Challenge
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`AdminLessonChallenges component Should refuse updating challenge with incorrect info 1`] = `
<div>
  <div
    class="mt-3"
  >
    <div
      class="row"
    >
      <div
        class="card shadow-sm col-12"
      >
        <div
          class="card-body text-center"
        >
          <h2
            class="card-title font-weight-bold mb-3"
          >
            Sum of 2 Numbers
          </h2>
        </div>
        <div
          class="text-left"
        >
          <div
            class="d-flex flex-column ml-3 mr-3 mb-4"
          >
            <h5
              data-testid="h5title1"
            >
              Title
            </h5>
            <input
              class="form-control"
              data-testid="input1"
              placeholder=""
              type="text"
              value="New title"
            />
          </div>
          <div
            class="d-flex flex-column ml-3 mr-3 mb-4"
          >
            <h5
              data-testid="h5description2"
            >
              Description
            </h5>
            <div
              style="background-color: white;"
            >
              <button
                class="btn"
                style="color: rgb(41, 41, 41);"
              >
                Write
              </button>
              <button
                class="btn"
                style="color: rgb(142, 142, 142);"
              >
                Preview
              </button>
              <textarea
                data-testid="textbox"
                placeholder="Type something..."
                style="padding: 1rem; width: 100%;"
              >
                New description
              </textarea>
            </div>
          </div>
          <div
            class="d-flex flex-column ml-3 mr-3 mb-4"
          >
            <h5
              data-testid="h5order3"
            >
              Order
            </h5>
            <input
              class="form-control"
              data-testid="input3"
              placeholder=""
              type="text"
              value=""
            />
            <h6
              class="text-danger mb-0"
            >
              Required
            </h6>
          </div>
        </div>
        <div
          class="text-center mb-4"
        >
          <button
            class="btn bg-primary"
            style="color: rgb(255, 255, 255);"
          >
            Update Challenge
          </button>
        </div>
      </div>
    </div>
  </div>
  <div
    class="mt-3"
  >
    <div
      class="row"
    >
      <div
        class="card shadow-sm col-12"
      >
        <div
          class="card-body text-center"
        >
          <h2
            class="card-title font-weight-bold mb-3"
          >
            Sum of 3 Numbers
          </h2>
        </div>
        <div
          class="text-left"
        >
          <div
            class="d-flex flex-column ml-3 mr-3 mb-4"
          >
            <h5
              data-testid="h5title1"
            >
              Title
            </h5>
            <input
              class="form-control"
              data-testid="input1"
              placeholder=""
              type="text"
              value="Sum of 3 Numbers"
            />
          </div>
          <div
            class="d-flex flex-column ml-3 mr-3 mb-4"
          >
            <h5
              data-testid="h5description2"
            >
              Description
            </h5>
            <div
              style="background-color: white;"
            >
              <button
                class="btn"
                style="color: rgb(41, 41, 41);"
              >
                Write
              </button>
              <button
                class="btn"
                style="color: rgb(142, 142, 142);"
              >
                Preview
              </button>
              <textarea
                data-testid="textbox"
                placeholder="Type something..."
                style="padding: 1rem; width: 100%;"
              >
                Write a function that takes in 3 numbers and returns their sum.

Here's how another developer might use your function:

\`\`\`
solution(5,9,2) // Should return 16
solution(4,1,9) // Should return 14
\`\`\`
              </textarea>
            </div>
          </div>
          <div
            class="d-flex flex-column ml-3 mr-3 mb-4"
          >
            <h5
              data-testid="h5order3"
            >
              Order
            </h5>
            <input
              class="form-control"
              data-testid="input3"
              placeholder=""
              type="text"
              value="2"
            />
          </div>
        </div>
        <div
          class="text-center mb-4"
        >
          <button
            class="btn bg-primary"
            style="color: rgb(255, 255, 255);"
          >
            Update Challenge
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`AdminLessonChallenges component Should render a challenge without title 1`] = `
<div>
  <div
    class="mt-3"
  >
    <div
      class="row"
    >
      <div
        class="card shadow-sm col-12"
      >
        <div
          class="card-body text-center"
        />
        <div
          class="text-left"
        >
          <div
            class="d-flex flex-column ml-3 mr-3 mb-4"
          >
            <h5
              data-testid="h5description1"
            >
              Description
            </h5>
            <div
              style="background-color: white;"
            >
              <button
                class="btn"
                style="color: rgb(41, 41, 41);"
              >
                Write
              </button>
              <button
                class="btn"
                style="color: rgb(142, 142, 142);"
              >
                Preview
              </button>
              <textarea
                data-testid="textbox"
                placeholder="Type something..."
                style="padding: 1rem; width: 100%;"
              >
                Write a function that takes in 2 numbers and returns their sum.

Here's how another developer might use your function:

\`\`\`
solution(5,9) // Should return 14
solution(4,1) // Should return 5
\`\`\`
              </textarea>
            </div>
          </div>
          <div
            class="d-flex flex-column ml-3 mr-3 mb-4"
          >
            <h5
              data-testid="h5order2"
            >
              Order
            </h5>
            <input
              class="form-control"
              data-testid="input2"
              placeholder=""
              type="text"
              value="1"
            />
          </div>
        </div>
        <div
          class="text-center mb-4"
        >
          <button
            class="btn bg-primary"
            style="color: rgb(255, 255, 255);"
          >
            Update Challenge
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`AdminLessonChallenges component Should render nulled challenges 1`] = `<div />`;

exports[`AdminLessonChallenges component Should render undefined challenges 1`] = `<div />`;

exports[`AdminLessonChallenges component Should update challenge 1`] = `
<div>
  <div
    class="mt-3"
  >
    <div
      class="row"
    >
      <div
        class="card shadow-sm col-12"
      >
        <div
          class="card-body text-center"
        >
          <h2
            class="card-title font-weight-bold mb-3"
          >
            Sum of 2 Numbers
          </h2>
        </div>
        <div
          class="text-left"
        >
          <div
            class="d-flex flex-column ml-3 mr-3 mb-4"
          >
            <h5
              data-testid="h5title1"
            >
              Title
            </h5>
            <input
              class="form-control"
              data-testid="input1"
              placeholder=""
              type="text"
              value="Sum of 2 Numbers"
            />
          </div>
          <div
            class="d-flex flex-column ml-3 mr-3 mb-4"
          >
            <h5
              data-testid="h5description2"
            >
              Description
            </h5>
            <div
              style="background-color: white;"
            >
              <button
                class="btn"
                style="color: rgb(41, 41, 41);"
              >
                Write
              </button>
              <button
                class="btn"
                style="color: rgb(142, 142, 142);"
              >
                Preview
              </button>
              <textarea
                data-testid="textbox"
                placeholder="Type something..."
                style="padding: 1rem; width: 100%;"
              >
                Write a function that takes in 2 numbers and returns their sum.

Here's how another developer might use your function:

\`\`\`
solution(5,9) // Should return 14
solution(4,1) // Should return 5
\`\`\`
              </textarea>
            </div>
          </div>
          <div
            class="d-flex flex-column ml-3 mr-3 mb-4"
          >
            <h5
              data-testid="h5order3"
            >
              Order
            </h5>
            <input
              class="form-control"
              data-testid="input3"
              placeholder=""
              type="text"
              value="1"
            />
          </div>
        </div>
        <div
          class="text-center mb-4"
        >
          <button
            class="btn bg-primary"
            style="color: rgb(255, 255, 255);"
          >
            Update Challenge
          </button>
        </div>
      </div>
    </div>
  </div>
  <div
    class="mt-3"
  >
    <div
      class="row"
    >
      <div
        class="card shadow-sm col-12"
      >
        <div
          class="card-body text-center"
        >
          <h2
            class="card-title font-weight-bold mb-3"
          >
            Sum of 3 Numbers
          </h2>
        </div>
        <div
          class="text-left"
        >
          <div
            class="d-flex flex-column ml-3 mr-3 mb-4"
          >
            <h5
              data-testid="h5title1"
            >
              Title
            </h5>
            <input
              class="form-control"
              data-testid="input1"
              placeholder=""
              type="text"
              value="Sum of 3 Numbers"
            />
          </div>
          <div
            class="d-flex flex-column ml-3 mr-3 mb-4"
          >
            <h5
              data-testid="h5description2"
            >
              Description
            </h5>
            <div
              style="background-color: white;"
            >
              <button
                class="btn"
                style="color: rgb(41, 41, 41);"
              >
                Write
              </button>
              <button
                class="btn"
                style="color: rgb(142, 142, 142);"
              >
                Preview
              </button>
              <textarea
                data-testid="textbox"
                placeholder="Type something..."
                style="padding: 1rem; width: 100%;"
              >
                Write a function that takes in 3 numbers and returns their sum.

Here's how another developer might use your function:

\`\`\`
solution(5,9,2) // Should return 16
solution(4,1,9) // Should return 14
\`\`\`
              </textarea>
            </div>
          </div>
          <div
            class="d-flex flex-column ml-3 mr-3 mb-4"
          >
            <h5
              data-testid="h5order3"
            >
              Order
            </h5>
            <input
              class="form-control"
              data-testid="input3"
              placeholder=""
              type="text"
              value="2"
            />
          </div>
        </div>
        <div
          class="text-center mb-4"
        >
          <button
            class="btn bg-primary"
            style="color: rgb(255, 255, 255);"
          >
            Update Challenge
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`AdminLessonChallenges component Should update challenge 2`] = `
<div>
  <div
    class="mt-3"
  >
    <div
      class="row"
    >
      <div
        class="card shadow-sm col-12"
      >
        <div
          class="card-body text-center"
        >
          <h2
            class="card-title font-weight-bold mb-3"
          >
            Sum of 2 Numbers
          </h2>
        </div>
        <div
          class="text-left"
        >
          <div
            class="d-flex flex-column ml-3 mr-3 mb-4"
          >
            <h5
              data-testid="h5title1"
            >
              Title
            </h5>
            <input
              class="form-control"
              data-testid="input1"
              placeholder=""
              type="text"
              value="New title"
            />
          </div>
          <div
            class="d-flex flex-column ml-3 mr-3 mb-4"
          >
            <h5
              data-testid="h5description2"
            >
              Description
            </h5>
            <div
              style="background-color: white;"
            >
              <button
                class="btn"
                style="color: rgb(41, 41, 41);"
              >
                Write
              </button>
              <button
                class="btn"
                style="color: rgb(142, 142, 142);"
              >
                Preview
              </button>
              <textarea
                data-testid="textbox"
                placeholder="Type something..."
                style="padding: 1rem; width: 100%;"
              >
                New description
              </textarea>
            </div>
          </div>
          <div
            class="d-flex flex-column ml-3 mr-3 mb-4"
          >
            <h5
              data-testid="h5order3"
            >
              Order
            </h5>
            <input
              class="form-control"
              data-testid="input3"
              placeholder=""
              type="text"
              value="15"
            />
          </div>
        </div>
        <div
          class="text-center mb-4"
        >
          <button
            class="btn bg-primary"
            style="color: rgb(255, 255, 255);"
          >
            Update Challenge
          </button>
        </div>
      </div>
    </div>
  </div>
  <div
    class="mt-3"
  >
    <div
      class="row"
    >
      <div
        class="card shadow-sm col-12"
      >
        <div
          class="card-body text-center"
        >
          <h2
            class="card-title font-weight-bold mb-3"
          >
            Sum of 3 Numbers
          </h2>
        </div>
        <div
          class="text-left"
        >
          <div
            class="d-flex flex-column ml-3 mr-3 mb-4"
          >
            <h5
              data-testid="h5title1"
            >
              Title
            </h5>
            <input
              class="form-control"
              data-testid="input1"
              placeholder=""
              type="text"
              value="Sum of 3 Numbers"
            />
          </div>
          <div
            class="d-flex flex-column ml-3 mr-3 mb-4"
          >
            <h5
              data-testid="h5description2"
            >
              Description
            </h5>
            <div
              style="background-color: white;"
            >
              <button
                class="btn"
                style="color: rgb(41, 41, 41);"
              >
                Write
              </button>
              <button
                class="btn"
                style="color: rgb(142, 142, 142);"
              >
                Preview
              </button>
              <textarea
                data-testid="textbox"
                placeholder="Type something..."
                style="padding: 1rem; width: 100%;"
              >
                Write a function that takes in 3 numbers and returns their sum.

Here's how another developer might use your function:

\`\`\`
solution(5,9,2) // Should return 16
solution(4,1,9) // Should return 14
\`\`\`
              </textarea>
            </div>
          </div>
          <div
            class="d-flex flex-column ml-3 mr-3 mb-4"
          >
            <h5
              data-testid="h5order3"
            >
              Order
            </h5>
            <input
              class="form-control"
              data-testid="input3"
              placeholder=""
              type="text"
              value="2"
            />
          </div>
        </div>
        <div
          class="text-center mb-4"
        >
          <button
            class="btn bg-primary"
            style="color: rgb(255, 255, 255);"
          >
            Update Challenge
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
